go/chat/localizer.go

```diff
@@ -0,0 +1,952 @@
+	"time"
+
+	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/chat/storage"
+	"github.com/keybase/client/go/chat/types"
+	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/gregor1"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/client/go/teams"
+	"github.com/keybase/client/go/uidmap"
+	"golang.org/x/sync/errgroup"
+)
+
+type conversationLocalizer interface {
+	Localize(ctx context.Context, uid gregor1.UID, inbox types.Inbox, maxLocalize *int) ([]chat1.ConversationLocal, error)
+	Name() string
+}
+
+type baseLocalizer struct {
+	globals.Contextified
+	utils.DebugLabeler
+	pipeline *localizerPipeline
+}
+
+func newBaseLocalizer(g *globals.Context, pipeline *localizerPipeline) *baseLocalizer {
+	return &baseLocalizer{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g.GetLog(), "baseLocalizer", false),
+		pipeline:     pipeline,
+	}
+}
+
+func (b *baseLocalizer) filterSelfFinalized(ctx context.Context, inbox types.Inbox) (res types.Inbox) {
+	username := b.G().Env.GetUsername().String()
+	res = inbox
+	res.ConvsUnverified = nil
+	for _, conv := range inbox.ConvsUnverified {
+		if conv.Conv.GetMembersType() == chat1.ConversationMembersType_KBFS &&
+			conv.Conv.GetFinalizeInfo() != nil &&
+			// If reset user is the current user, or is blank (only way such a thing could be in our
+			// inbox is if the current user is the one that reset)
+			(conv.Conv.GetFinalizeInfo().ResetUser == username ||
+				conv.Conv.GetFinalizeInfo().ResetUser == "") {
+			b.Debug(ctx, "baseLocalizer: skipping own finalized convo: %s name: %s", conv.GetConvID())
+			continue
+		}
+		res.ConvsUnverified = append(res.ConvsUnverified, conv)
+	}
+	return res
+}
+
+func (b *baseLocalizer) getConvs(inbox types.Inbox, maxLocalize *int) []chat1.Conversation {
+	convs := utils.PluckConvs(inbox.ConvsUnverified)
+	if maxLocalize == nil {
+		return convs
+	}
+	return convs[:*maxLocalize]
+}
+
+type blockingLocalizer struct {
+	globals.Contextified
+	utils.DebugLabeler
+	*baseLocalizer
+
+	localizeCb chan types.AsyncInboxResult
+}
+
+func newBlockingLocalizer(g *globals.Context, pipeline *localizerPipeline,
+	localizeCb chan types.AsyncInboxResult) *blockingLocalizer {
+	return &blockingLocalizer{
+		Contextified:  globals.NewContextified(g),
+		DebugLabeler:  utils.NewDebugLabeler(g.GetLog(), "blockingLocalizer", false),
+		baseLocalizer: newBaseLocalizer(g, pipeline),
+		localizeCb:    localizeCb,
+	}
+}
+
+func (b *blockingLocalizer) Localize(ctx context.Context, uid gregor1.UID, inbox types.Inbox,
+	maxLocalize *int) (res []chat1.ConversationLocal, err error) {
+	defer b.Trace(ctx, func() error { return err }, "Localize")()
+	inbox = b.filterSelfFinalized(ctx, inbox)
+	convs := b.getConvs(inbox, maxLocalize)
+	b.baseLocalizer.pipeline.queue(ctx, uid, convs, b.localizeCb)
```

Comment:
should this be returning an error on the queue being full? we check `err` below but it's from the return value